File: server/TerraformExplorer/Services/AccountService.cs
using TerraformExplorer.Exceptions;
using TerraformExplorer.Models;
using TerraformExplorer.Models.Requests;
using TerraformExplorer.Models.Responses;
using TerraformExplorer.Settings;
using TerraformExplorer.Utils;

namespace TerraformExplorer.Services;

public class AccountService
{
    private readonly TerraformService _terraformService;
    private readonly FileSystemService _fileSystemService;
    private readonly TerraformSettings _terraformSettings;
    // private readonly AwsS3Service _s3Service;

    public AccountService(
        TerraformService terraformService, 
        FileSystemService fileSystemService, 
        // AwsS3Service s3Service,
        TerraformSettings terraformSettings)
    {
        _terraformService = terraformService;
        _fileSystemService = fileSystemService;
        // _s3Service = s3Service;
        _terraformSettings = terraformSettings;
    }
    
    public void CopyAccount(AccountCopyRequest request)
    {
        var sourcePath = GetPath(request.SourceAccountName);
        var destPath = GetPath(request.DestinationAccountName);

        _fileSystemService.CopyDirectory(sourcePath, destPath);
    }
    
    private string GetPath(string name)
    {
        return Path.Combine(_terraformSettings.GetAccountsPath(), name);
    }

    public List<ItemResponse> List()
    {
        var accounts = _terraformService.GetStructure().Accounts;
        return accounts.Select(account => new ItemResponse { Name = account.Name, Path = account.Path }).ToList();
    }

    public ItemResponse Get(string name)
    {
        var path = GetPath(name);
        _fileSystemService.EnsureExists(path, "Conta");
        return new ItemResponse { Name = name, Path = path };
    }

    public void Create(string name)
    {
        var path = GetPath(name);
        _fileSystemService.CreateDirectory(path);
    }

    public void Delete(string name)
    {
        var path = GetPath(name);
        _fileSystemService.DeleteDirectory(path);
    }
    
    public void Rename(string oldName, string newName)
    {
        var oldPath = GetPath(oldName);
        var newPath = GetPath(newName);
        _fileSystemService.Rename(oldPath, newPath);
    }
    
    public async Task LinkProviderToAccount(string accountName, SetAwsConfigRequest request)
    {
        var structure = TerraformStructureLoader.Load(_terraformSettings);
        var account = structure.Accounts.FirstOrDefault(a => a.Name == accountName)
                      ?? throw new NotFoundException($"Conta {accountName} não encontrada.");
        
        account.AwsProfile = request.Profile;
        account.AssumeRoleArn = request.RoleArn;
        account.Region = request.Region;
        
        ProviderTfGenerator.GenerateProvider(account);
        
        var credentialsPath = Path.Combine(_terraformSettings.GetProvidersPath(), "credentials");
        
        var s3Service = new AwsS3Service(
            account.AwsProfile,
            credentialsPath,
            account.AssumeRoleArn,
            account.Region
        );
        await ProviderTfGenerator.GenerateBackendAsync(account, s3Service);
    }
    
    private string GetConfigPath()
    {   
        return Path.Combine(_terraformSettings.GetProvidersPath(), "config");
    }
    
}

File: apps/web/src/services/AccountService.ts
import HttpClient from "@/clients/HttpClient";
import type { AccountCopyRequest } from "@/interfaces/requests/AccountCopyRequest";
import type { CreateItemRequest } from "@/interfaces/requests/CreateItemRequest";
import type { RenameRequest } from "@/interfaces/requests/RenameRequest";
import type { SetAwsConfigRequest } from "@/interfaces/requests/SetAwsConfigRequest";
import type { ItemResponse } from "@/interfaces/responses/ItemResponse";
import type { SuccessResponse } from "@/interfaces/responses/SuccessResponse";

export class AccountService {
  private static httpClient: HttpClient;
  private static baseURL: string = `${import.meta.env.VITE_API_BASE_URL}/account`;

  private static getHttpClient(): HttpClient {
    if (!this.httpClient) {
      this.httpClient = new HttpClient(this.baseURL);
    }
    return this.httpClient;
  }

  static async list(): Promise<ItemResponse[]> {
    return this.getHttpClient().get<ItemResponse[]>("");
  }

  static async get(name: string): Promise<ItemResponse> {
    return this.getHttpClient().get<ItemResponse>(`/${encodeURIComponent(name)}`);
  }

  static async create(request: CreateItemRequest): Promise<SuccessResponse> {
    return this.getHttpClient().post<SuccessResponse>("", request);
  }

  static async delete(name: string): Promise<SuccessResponse> {
    return this.getHttpClient().delete<SuccessResponse>(`/${encodeURIComponent(name)}`);
  }

  static async rename(name: string, req: RenameRequest): Promise<SuccessResponse> {
    return this.getHttpClient().put<SuccessResponse>(`/${encodeURIComponent(name)}`, req);
  }

  static async copyAccount(request: AccountCopyRequest): Promise<SuccessResponse> {
    return this.getHttpClient().post<SuccessResponse>("copy", request);
  }

  static async linkProviderToAccount(accountName: string, request: SetAwsConfigRequest): Promise<SuccessResponse> {
    return this.getHttpClient().post<SuccessResponse>(`${encodeURIComponent(accountName)}/link-provider-to-account`, request);
  }
}